pkg/logql/rangemapper.go

```diff
@@ -327,6 +340,74 @@ func (m RangeMapper) getOriginalOffset(expr syntax.SampleExpr) (offset time.Dura
 // rangeInterval should be greater than m.splitByInterval, otherwise the resultant expression
 // will have an unnecessary aggregation operation
 func (m RangeMapper) mapConcatSampleExpr(expr syntax.SampleExpr, rangeInterval time.Duration, recorder *downstreamRecorder) syntax.SampleExpr {
+	if m.splitAlignTs.IsZero() {
+		return m.rangeSplit(expr, rangeInterval, recorder)
+	}
+	return m.rangeSplitAlign(expr, rangeInterval, recorder)
+}
+
+// rangeSplitAlign try to split given `rangeInterval` into units of `m.splitByInterval` by making sure `rangeInterval` is aligned with `m.splitByInterval` for as much as the units as possible.
+// Consider following example with real use case.
+// Instant Query: `sum(rate({foo="bar"}[3h])`
+// execTs: 12:34:00
+// splitBy: 1h
+// Given above parameters, queries will be split into following
+// 1. sum(rate({foo="bar"}[34m]))
+// 2. sum(rate({foo="bar"}[1h] offset 34m))
+// 3. sum(rate({foo="bar"}[1h] offset 1h34m))
+// 4. sum(rate({foo="bar"}[26m] offset 2h34m))
+// All with execTs: 12:34:00. Here (2) and (3) range values (1h) aligned with splitByInteval (1h).
+func (m RangeMapper) rangeSplitAlign(
+	expr syntax.SampleExpr, rangeInterval time.Duration, recorder *downstreamRecorder,
+) syntax.SampleExpr {
+	if rangeInterval <= m.splitByInterval {
+		return expr
+	}
+
+	originalOffset, err := m.getOriginalOffset(expr)
+	if err != nil {
+		return expr
+	}
+
+	splits := int(math.Ceil(float64(rangeInterval) / float64(m.splitByInterval))) // without first aligned subquery
+	align := m.splitAlignTs.Sub(m.splitAlignTs.Truncate(m.splitByInterval))       // say, 12:34:00 - 12:00:00(truncated) = 34m
+
+	if align == 0 {
+		return m.rangeSplit(expr, rangeInterval, recorder) // Don't have to align
+	}
+
+	splits += 1
+
+	var (
+		newRng      time.Duration = align
+		newOffset   time.Duration = originalOffset
+		downstreams *ConcatSampleExpr
+	)
+
+	// first subquery
+	downstreams = appendDownstream(downstreams, expr, newRng, newOffset)
+
+	newOffset += align         // e.g: offset 34m
+	newRng = m.splitByInterval // [1h]
+
+	// Rest of the subqueries. Always aligned with splitBy
+	for i := 0; i < splits-2; i++ {
+		downstreams = appendDownstream(downstreams, expr, newRng, newOffset)
+		newOffset += m.splitByInterval
+	}
+
+	// last subquery
+	newRng = m.splitByInterval - align // e.g: [24m]
```

Comment:
could this be 0 or neg for the last split? thinking if we will always have splits+1 when the query is unaligned.
what do you think about generating sub-queries as long as rangeInterval > 0 decreasing it with each generated query.

1. reduce rangeInterval by `align` for the first query
2. keep adding sub-queries that are aligned to splitInterval and sub `splitByInterval` from `rangeInterval`
3. when the left over `rangeInterval` is smaller than `splitByInterval`, use it to generate the last query

Reply:
> could this be 0 or neg for the last split?

Failing to understand how this could be possible. This is based on the invariant that `splitByInterval` will **always be greater than** `align`.

If you see how we calculate `align` in the first place, 
```
align := m.splitAlignTs.Sub(m.splitAlignTs.Truncate(m.splitByInterval))       // say, 12:34:00 - 12:00:00(truncated) = 34m
```
wondering at what case, this `align` can be greater than or equal to `splitbyInterval`.  Am I missing anything here?

Also, the approach you explained focus only on the `rangeInterval` but not on the `offset` we generate for each subqueries. 

So basically we don't subtract `splitByInterval` from `rangeInterval` for all subqueries, but instead we adjust the offset. keeping `rangeInterval` aligned with `splitByInterval` (except first and last subquery). 

So once we remove offset and adjust start and end time we have sub queries that have aligned both `rangeInterval` and `execTs`.

May be I'm missing something that you are trying to explain?

Reply:
taking the same example from the comments, if we take the case where orig query uses a  range of 2h34m and executes at 12:34:00 `sum(rate({foo="bar"}[2h34m])`

these would be the expected sub queries:
1. sum(rate({foo="bar"}[34m]))
2. sum(rate({foo="bar"}[1h] offset 34m))
3. sum(rate({foo="bar"}[1h] offset 1h34m))

4th sub-query would not be need in this case. don't think it could be a neg value, but 0 is a possibility but a very unlikely one. we could just as well solve this with a check to see if the left over range is gt 0.

Reply:
> 4th sub-query would not be need in this case. don't think it could be a neg value, but 0 is a possibility but a very unlikely one. we could just as well solve this with a check to see if the left over range is gt 0.

ACK that. I cleanup up the method to address this part 👍